cvmfs/session_context.cc

```diff
@@ -299,8 +294,8 @@ bool SessionContext::InitializeDerived(uint64_t max_queue_size) {
 }
 
 bool SessionContext::FinalizeDerived() {
-  if (atomic_cas32(&worker_terminate_, 0, 1))
-    pthread_join(worker_, NULL);
+  upload_jobs_->Enqueue(&terminator_);
```

Comment:
```suggestion
  // Note: in FinalizedDerived, we know that the worker is running. The SessionContext is only called from
  // GatewayUploader::FinalizeSession(), which in turn is from Spooler::FinalizeSession().  The Spooler ensures that
  // GatewayUploader::Initialize() is called on construction.
  // TODO(jblomer): Refactor SessionContext (and Uploader*) classes to use a factory method for construction.
  upload_jobs_->Enqueue(&terminator_);
```

Reply:
Thank you very much - have added this to the commit